nebula_decoders/include/nebula_decoders/nebula_decoders_velodyne/decoders/velodyne_generic_decoder.hpp

```diff
@@ -0,0 +1,445 @@
+      rotation_radians_[rot_index] = rotation;
+      cos_rot_table_[rot_index] = cosf(rotation);
+      sin_rot_table_[rot_index] = sinf(rotation);
+    }
+
+    phase_ = (uint16_t)round(sensor_configuration_->scan_phase * 100);
+
+    // fill vls128 laser azimuth cache
+    sensor_.fillAzimuthCache();  // TODO: predefine azimuth cache and remove this for performance?
+
+    // timing table calculation, from velodyne user manual p.64
+    timing_offsets_.resize(BLOCKS_PER_PACKET);  // x dir size
+    for (size_t i = 0; i < timing_offsets_.size(); ++i) {
+      timing_offsets_[i].resize(
+        CHANNELS_PER_BLOCK + SensorT::num_maintenance_periods);  // y dir size
+    }
+
+    double full_firing_cycle_s = SensorT::full_firing_cycle_s;
+    double single_firing_s = SensorT::single_firing_s;
+    double offset_packet_time = SensorT::offset_packet_time;
+    bool dual_mode = sensor_configuration_->return_mode == ReturnMode::DUAL;
+    double firing_sequence_index, data_point_index;
+    // compute timing offsets
+    for (size_t x = 0; x < timing_offsets_.size(); ++x) {
+      for (size_t y = 0; y < timing_offsets_[x].size(); ++y) {
+        if (dual_mode) {
+          firing_sequence_index = static_cast<long>(x * SensorT::firing_sequences_per_block / 2) +
+                                  (y / SensorT::channels_per_firing_sequence);
+        } else {
+          firing_sequence_index =
+            static_cast<long>(x * SensorT::firing_sequences_per_block) +
+            (y / SensorT::channels_per_firing_sequence);  // cast to long to make double
+                                                          // multiplication integer division
+        }
+        data_point_index =
+          (y % SensorT::channels_per_firing_sequence) / SensorT::num_simultaneous_firings -
+          offset_packet_time;
+        timing_offsets_[x][y] =
+          (full_firing_cycle_s * firing_sequence_index) + (single_firing_s * data_point_index);
+      }
+    }
+  }
+
+  // DONE
+  bool hasScanned() { return has_scanned_; }
+
+  std::tuple<drivers::NebulaPointCloudPtr, double> get_pointcloud()
+  {
+    double phase = angles::from_degrees(sensor_configuration_->scan_phase);
+    if (!scan_pc_->points.empty()) {
+      auto current_azimuth = scan_pc_->points.back().azimuth;
+      auto phase_diff = (2 * M_PI + current_azimuth - phase);
+      while (phase_diff < M_PI_2 && !scan_pc_->points.empty()) {
+        overflow_pc_->points.push_back(scan_pc_->points.back());
+        scan_pc_->points.pop_back();
+        current_azimuth = scan_pc_->points.back().azimuth;
+        phase_diff = (2 * M_PI + current_azimuth - phase);
+      }
+      overflow_pc_->width = overflow_pc_->points.size();
+      scan_pc_->width = scan_pc_->points.size();
+      scan_pc_->height = 1;
+    }
+    return std::make_tuple(scan_pc_, scan_timestamp_);
+  }
+
+  // DONE
+  int pointsPerPacket() { return BLOCKS_PER_PACKET * CHANNELS_PER_BLOCK; }
```

Comment:
Are these constants the same for all Velodyne sensors? Ideally these would be in `VelodyneSensor` or in its subclasses.